src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -189,13 +189,55 @@ private NamedTypeSymbol ConstructAnonymousDelegateImplementationSymbol(Anonymous
             // If all parameter types and return type are valid type arguments, construct
             // the delegate type from a generic template. Otherwise, use a non-generic template.
             bool useUpdatedEscapeRules = Compilation.SourceModule.UseUpdatedEscapeRules;
-            if (allValidTypeArguments(useUpdatedEscapeRules, typeDescr))
+            if (allValidTypeArguments(useUpdatedEscapeRules, typeDescr, out var needsIndexedName))
             {
                 var fields = typeDescr.Fields;
                 Debug.Assert(fields.All(f => hasDefaultScope(useUpdatedEscapeRules, f)));
 
-                bool returnsVoid = fields.Last().Type.IsVoidType();
+                bool returnsVoid = fields[^1].Type.IsVoidType();
                 int nTypeArguments = fields.Length - (returnsVoid ? 1 : 0);
+                var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(nTypeArguments);
+                for (int i = 0; i < nTypeArguments; i++)
+                {
+                    var field = fields[i];
+                    if (field.IsParams)
+                    {
+                        Debug.Assert(needsIndexedName);
+                        Debug.Assert(i == fields.Length - 2);
+                        Debug.Assert(field.Type.IsSZArray());
+                        typeArgumentsBuilder.Add(((ArrayTypeSymbol)field.Type).ElementTypeWithAnnotations);
+                    }
+                    else
+                    {
+                        typeArgumentsBuilder.Add(field.TypeWithAnnotations);
+                    }
+                }
+
+                var typeArguments = typeArgumentsBuilder.ToImmutableAndFree();
+
+                // Delegate name cannot be fully determined by its signature (e.g., it has default parameter values).
+                if (needsIndexedName)
+                {
+                    Debug.Assert(nTypeArguments != 0);
+
+                    // Construct key for the delegate with type parameters.
+                    var genericFieldTypes = IndexedTypeParameterSymbol.Take(nTypeArguments);
```

Comment:
>var genericFieldTypes = IndexedTypeParameterSymbol.Take(nTypeArguments); [](http://example.com/codeflow?start=20&length=72)

It feels like we are missing array creation for the `params` field

Reply:
Not really, this creates `genericTypeDescr` that's only used as a key to `AnonymousDelegates` dictionary (that's why I changed the comment to reflect that). So yes, the `AnonymousTypeDescriptor` will have generic `params` non-array parameter, technically incorrect, but not used anywhere, so I thought it's unnecessary to replace `T` by `T[]` here.

Reply:
> Maybe I should extend the comment to explain this?

I think the key is supposed to be an accurate reflection of delegate's signature where indexed type parameters represent future delegate type parameters. You might argue, that today due to different reasons it is not possible to have a key collision between delegates of different signatures. However, I do not think that taking this chance is worth the risk. Making an array type symbol isn't that complicated and the code will be more robust in the long term.

Reply:
There is a helper ``` ArrayTypeSymbol.WithElementType``` that can be used on the original array type

Reply:
OK, makes sense, thanks. Changed.